File: dotnet/packages/Microsoft.TeamsAI/Microsoft.TeamsAI/Application/Meetings/Meetings.cs
using Microsoft.Bot.Connector;
using Microsoft.Bot.Schema;
using Microsoft.Bot.Schema.Teams;
using Microsoft.Teams.AI.State;
using Microsoft.Teams.AI.Utilities;

namespace Microsoft.Teams.AI
{
    /// <summary>
    /// Meetings class to enable fluent style registration of handlers related to Microsoft Teams Meetings.
    /// </summary>
    /// <typeparam name="TState">The type of the turn state object used by the application.</typeparam>
    public class Meetings<TState>
        where TState : TurnState<Record, Record, TempState>, new()
    {
        private readonly Application<TState> _app;

        /// <summary>
        /// Creates a new instance of the Meetings class.
        /// </summary>
        /// <param name="app"></param> The top level application class to register handlers with.
        public Meetings(Application<TState> app)
        {
            this._app = app;
        }

        /// <summary>
        /// Handles Microsoft Teams meeting start events.
        /// </summary>
        /// <param name="handler">Function to call when a Microsoft Teams meeting start event activity is received from the connector.</param>
        /// <returns>The application instance for chaining purposes.</returns>
        public Application<TState> OnStart(MeetingStartHandler<TState> handler)
        {
            Verify.ParamNotNull(handler);
            RouteSelector routeSelector = (context, _) => Task.FromResult
            (
                string.Equals(context.Activity?.Type, ActivityTypes.Event, StringComparison.OrdinalIgnoreCase)
                && string.Equals(context.Activity?.ChannelId, Channels.Msteams)
                && string.Equals(context.Activity?.Name, "application/vnd.microsoft.meetingStart")
            );
            RouteHandler<TState> routeHandler = async (turnContext, turnState, cancellationToken) =>
            {
                MeetingStartEventDetails meeting = ActivityUtilities.GetTypedValue<MeetingStartEventDetails>(turnContext.Activity) ?? new();
                await handler(turnContext, turnState, meeting, cancellationToken);
            };
            _app.AddRoute(routeSelector, routeHandler, isInvokeRoute: false);
            return _app;
        }

        /// <summary>
        /// Handles Microsoft Teams meeting end events.
        /// </summary>
        /// <param name="handler">Function to call when a Microsoft Teams meeting end event activity is received from the connector.</param>
        /// <returns>The application instance for chaining purposes.</returns>
        public Application<TState> OnEnd(MeetingEndHandler<TState> handler)
        {
            Verify.ParamNotNull(handler);
            RouteSelector routeSelector = (context, _) => Task.FromResult
            (
                string.Equals(context.Activity?.Type, ActivityTypes.Event, StringComparison.OrdinalIgnoreCase)
                && string.Equals(context.Activity?.ChannelId, Channels.Msteams)
                && string.Equals(context.Activity?.Name, "application/vnd.microsoft.meetingEnd")
            );
            RouteHandler<TState> routeHandler = async (turnContext, turnState, cancellationToken) =>
            {
                MeetingEndEventDetails meeting = ActivityUtilities.GetTypedValue<MeetingEndEventDetails>(turnContext.Activity) ?? new();
                await handler(turnContext, turnState, meeting, cancellationToken);
            };
            _app.AddRoute(routeSelector, routeHandler, isInvokeRoute: false);
            return _app;
        }

        /// <summary>
        /// Handles Microsoft Teams meeting participants join events.
        /// </summary>
        /// <param name="handler">Function to call when a Microsoft Teams meeting participants join event activity is received from the connector.</param>
        /// <returns>The application instance for chaining purposes.</returns>
        public Application<TState> OnParticipantsJoin(MeetingParticipantsEventHandler<TState> handler)
        {
            Verify.ParamNotNull(handler);
            RouteSelector routeSelector = (context, _) => Task.FromResult
            (
                string.Equals(context.Activity?.Type, ActivityTypes.Event, StringComparison.OrdinalIgnoreCase)
                && string.Equals(context.Activity?.ChannelId, Channels.Msteams)
                && string.Equals(context.Activity?.Name, "application/vnd.microsoft.meetingParticipantJoin")
            );
            RouteHandler<TState> routeHandler = async (turnContext, turnState, cancellationToken) =>
            {
                MeetingParticipantsEventDetails meeting = ActivityUtilities.GetTypedValue<MeetingParticipantsEventDetails>(turnContext.Activity) ?? new();
                await handler(turnContext, turnState, meeting, cancellationToken);
            };
            _app.AddRoute(routeSelector, routeHandler, isInvokeRoute: false);
            return _app;
        }

        /// <summary>
        /// Handles Microsoft Teams meeting participants leave events.
        /// </summary>
        /// <param name="handler">Function to call when a Microsoft Teams meeting participants leave event activity is received from the connector.</param>
        /// <returns>The application instance for chaining purposes.</returns>
        public Application<TState> OnParticipantsLeave(MeetingParticipantsEventHandler<TState> handler)
        {
            Verify.ParamNotNull(handler);
            RouteSelector routeSelector = (context, _) => Task.FromResult
            (
                string.Equals(context.Activity?.Type, ActivityTypes.Event, StringComparison.OrdinalIgnoreCase)
                && string.Equals(context.Activity?.ChannelId, Channels.Msteams)
                && string.Equals(context.Activity?.Name, "application/vnd.microsoft.meetingParticipantLeave")
            );
            RouteHandler<TState> routeHandler = async (turnContext, turnState, cancellationToken) =>
            {
                MeetingParticipantsEventDetails meeting = ActivityUtilities.GetTypedValue<MeetingParticipantsEventDetails>(turnContext.Activity) ?? new();
                await handler(turnContext, turnState, meeting, cancellationToken);
            };
            _app.AddRoute(routeSelector, routeHandler, isInvokeRoute: false);
            return _app;
        }
    }
}

File: js/packages/teams-ai/src/Meetings.ts
import {
    ActivityTypes,
    MeetingEndEventDetails,
    MeetingParticipantsEventDetails,
    MeetingStartEventDetails,
    TurnContext
} from 'botbuilder';
import { Application } from './Application';
import { TurnState } from './TurnState';

/**
 * Provides a set of methods for handling Teams meeting events.
 */
export class Meetings<TState extends TurnState = TurnState> {
    private readonly _app: Application<TState>;

    public constructor(app: Application<TState>) {
        this._app = app;
    }

    /**
     * Handles meeting start events for Microsoft Teams.
     * @template TState
     * @param {Function} handler - Function to call when the handler is triggered.
     * @returns {Application<TState>} The application for chaining purposes.
     */
    public start(
        handler: (context: TurnContext, state: TState, meeting: MeetingStartEventDetails) => Promise<void>
    ): Application<TState> {
        const selector = (context: TurnContext): Promise<boolean> => {
            return Promise.resolve(
                context.activity.type === ActivityTypes.Event &&
                    context.activity.channelId === 'msteams' &&
                    context.activity.name === 'application/vnd.microsoft.meetingStart'
            );
        };

        const handlerWrapper = (context: TurnContext, state: TState): Promise<void> => {
            const meeting = context.activity.value as MeetingStartEventDetails;
            return handler(context, state, meeting);
        };

        this._app.addRoute(selector, handlerWrapper);

        return this._app;
    }

    /**
     * Handles meeting end events for Microsoft Teams.
     * @template TState - The type of TurnState
     * @param {Function} handler - Function to call when the handler is triggered.
     * @returns {Application<TState>} The application for chaining purposes.
     */
    public end(
        handler: (context: TurnContext, state: TState, meeting: MeetingEndEventDetails) => Promise<void>
    ): Application<TState> {
        const selector = (context: TurnContext): Promise<boolean> => {
            return Promise.resolve(
                context.activity.type === ActivityTypes.Event &&
                    context.activity.channelId === 'msteams' &&
                    context.activity.name === 'application/vnd.microsoft.meetingEnd'
            );
        };

        const handlerWrapper = (context: TurnContext, state: TState): Promise<void> => {
            const meeting = context.activity.value as MeetingEndEventDetails;
            return handler(context, state, meeting);
        };

        this._app.addRoute(selector, handlerWrapper);

        return this._app;
    }

    /**
     * Handles meeting participant join events for Microsoft Teams.
     * @template TState
     * @param {Function} handler - Function to call when the handler is triggered.
     * @returns {Application<TState>} The application for chaining purposes.
     */
    public participantsJoin(
        handler: (context: TurnContext, state: TState, meeting: MeetingParticipantsEventDetails) => Promise<void>
    ): Application<TState> {
        const selector = (context: TurnContext): Promise<boolean> => {
            return Promise.resolve(
                context.activity.type === ActivityTypes.Event &&
                    context.activity.channelId === 'msteams' &&
                    context.activity.name === 'application/vnd.microsoft.meetingParticipantsJoin'
            );
        };

        const handlerWrapper = (context: TurnContext, state: TState): Promise<void> => {
            const meeting = context.activity.value as MeetingParticipantsEventDetails;
            return handler(context, state, meeting);
        };

        this._app.addRoute(selector, handlerWrapper);

        return this._app;
    }

    /**
     * Handles meeting participant leave events for Microsoft Teams.
     * @template TState - The type of TurnState
     * @param {Function} handler - Function to call when the handler is triggered.
     * @returns {Application<TState>} The application for chaining purposes.
     */
    public participantsLeave(
        handler: (context: TurnContext, state: TState, meeting: MeetingParticipantsEventDetails) => Promise<void>
    ): Application<TState> {
        const selector = (context: TurnContext): Promise<boolean> => {
            return Promise.resolve(
                context.activity.type === ActivityTypes.Event &&
                    context.activity.channelId === 'msteams' &&
                    context.activity.name === 'application/vnd.microsoft.meetingParticipantsLeave'
            );
        };

        const handlerWrapper = (context: TurnContext, state: TState): Promise<void> => {
            const meeting = context.activity.value as MeetingParticipantsEventDetails;
            return handler(context, state, meeting);
        };

        this._app.addRoute(selector, handlerWrapper);

        return this._app;
    }
}
